shadowsocks-csharp/Util/SystemProxy/Sysproxy.cs

```diff
@@ -82,41 +83,78 @@ public static void SetIEProxy(bool enable, bool global, string proxyServer, stri
         {
-            using (var process = new Process())
+            // using event to avoid hanging when redirect standard output/error
+            // ref: https://stackoverflow.com/questions/139593/processstartinfo-hanging-on-waitforexit-why
+            // and http://blog.csdn.net/zhangweixing0/article/details/7356841
+            using (AutoResetEvent outputWaitHandle = new AutoResetEvent(false))
+            using (AutoResetEvent errorWaitHandle = new AutoResetEvent(false))
             {
-                // Configure the process using the StartInfo properties.
-                process.StartInfo.FileName = Utils.GetTempPath("sysproxy.exe");
-                process.StartInfo.Arguments = arguments;
-                process.StartInfo.WorkingDirectory = Utils.GetTempPath();
-                process.StartInfo.WindowStyle = ProcessWindowStyle.Hidden;
-                process.StartInfo.UseShellExecute = false;
-                process.StartInfo.RedirectStandardError = true;
-                process.StartInfo.RedirectStandardOutput = true;
-
-                // Need to provide encoding info, or output/error strings we got will be wrong.
-                process.StartInfo.StandardOutputEncoding = Encoding.Unicode;
-                process.StartInfo.StandardErrorEncoding = Encoding.Unicode;
-
-                process.StartInfo.CreateNoWindow = true;
-                process.Start();
-
-                var stderr = process.StandardError.ReadToEnd();
-                var stdout = process.StandardOutput.ReadToEnd();
-
-                process.WaitForExit();
-
-                var exitCode = process.ExitCode;
-                if (exitCode != (int)RET_ERRORS.RET_NO_ERROR)
+                using (var process = new Process())
                 {
-                    throw new ProxyException(stderr);
-                }
+                    // Configure the process using the StartInfo properties.
+                    process.StartInfo.FileName = Utils.GetTempPath("sysproxy.exe");
+                    process.StartInfo.Arguments = arguments;
+                    process.StartInfo.WorkingDirectory = Utils.GetTempPath();
+                    process.StartInfo.WindowStyle = ProcessWindowStyle.Hidden;
+                    process.StartInfo.UseShellExecute = false;
+                    process.StartInfo.RedirectStandardError = true;
+                    process.StartInfo.RedirectStandardOutput = true;
+
+                    // Need to provide encoding info, or output/error strings we got will be wrong.
+                    process.StartInfo.StandardOutputEncoding = Encoding.Unicode;
+                    process.StartInfo.StandardErrorEncoding = Encoding.Unicode;
+
+                    process.StartInfo.CreateNoWindow = true;
+
+                    StringBuilder output = new StringBuilder();
+                    StringBuilder error = new StringBuilder();
+
+                    process.OutputDataReceived += (sender, e) =>

```

Comment:
Potential memory leak (not sure)?

Reply:
Potentially memory leak may occured cause the publisher(say process) lives longer than the subscriber(say lamda). In our case, the `process` is the only one that hold the reference of the `lamda` and will be disposed when it goes out of the `using` statment. So it's ok i think.

ref:
https://stackoverflow.com/questions/4526829/why-and-how-to-avoid-event-handler-memory-leaks